gossip/client/hubclient.go

```diff
@@ -0,0 +1,232 @@
+	// may have used a different signature from the same key, as some signature mechanisms
+	// (e.g. ECDSA) are not deterministic.
+	return &sgt, nil
+}
+
+// GetSTH retrieves the current STH from the hub.
+func (c *HubClient) GetSTH(ctx context.Context) (*api.SignedHubTreeHead, error) {
+	var rsp api.GetSTHResponse
+	httpRsp, body, err := c.GetAndParse(ctx, api.PathPrefix+api.GetSTHPath, nil, &rsp)
+	if err != nil {
+		return nil, err
+	}
+
+	// Pre-build a RspError for all error cases below.
+	rspErr := jsonclient.RspError{StatusCode: httpRsp.StatusCode, Body: body}
+
+	// First check the signature if we can.
+	if err := c.VerifySignature(rsp.TreeHeadData, rsp.HubSignature); err != nil {
+		rspErr.Err = err
+		return nil, rspErr
+	}
+
+	// Now it's safe to decode the TLS-encoded contents.
+	shth := api.SignedHubTreeHead{HubSignature: rsp.HubSignature}
+	if rest, err := tls.Unmarshal(rsp.TreeHeadData, &shth.TreeHead); err != nil {
+		rspErr.Err = err
+		return nil, rspErr
+	} else if len(rest) > 0 {
+		rspErr.Err = fmt.Errorf("trailing data (%d bytes) after TreeHead", len(rest))
+		return nil, rspErr
+	}
+	return &shth, nil
+}
+
+// VerifySignature checks the signature in sth, returning any error encountered or nil if verification is
+// successful.
+func (c *HubClient) VerifySignature(data, sig []byte) error {
+	if c.Verifier == nil {
+		// Can't verify signatures without a verifier
+		return nil
+	}
+	return tcrypto.Verify(c.Verifier.PubKey, c.Hash, data, sig)
+}
+
+// GetSTHConsistency retrieves the consistency proof between two hub tree heads.
+func (c *HubClient) GetSTHConsistency(ctx context.Context, first, second uint64) ([][]byte, error) {
+	if second < first {
+		return nil, fmt.Errorf("range inverted: second (%d) < first (%d)", second, first)
+	}
+	params := map[string]string{
+		api.GetSTHConsistencyFirst:  strconv.FormatUint(first, base10),
+		api.GetSTHConsistencySecond: strconv.FormatUint(second, base10),
+	}
+	var rsp api.GetSTHConsistencyResponse
+	_, _, err := c.GetAndParse(ctx, api.PathPrefix+api.GetSTHConsistencyPath, params, &rsp)
+	if err != nil {
+		return nil, err
+	}
+	return rsp.Consistency, nil
+}
+
+// GetProofByHash returns an audit path for the hash of a timestamped entry in the hub
+// at a particular tree size.
+func (c *HubClient) GetProofByHash(ctx context.Context, hash []byte, treeSize uint64) (*api.GetProofByHashResponse, error) {
+	params := map[string]string{
+		api.GetProofByHashSize: strconv.FormatUint(treeSize, base10),
+		api.GetProofByHashArg:  base64.StdEncoding.EncodeToString(hash),
+	}
+	var rsp api.GetProofByHashResponse
+	_, _, err := c.GetAndParse(ctx, api.PathPrefix+api.GetProofByHashPath, params, &rsp)
+	if err != nil {
+		return nil, err
+	}
+	return &rsp, nil
+}
+
+func (c *HubClient) getEntriesRsp(ctx context.Context, start, end int64) (int, []byte, [][]byte, error) {
+	if start < 0 {
+		return -1, nil, nil, errors.New("start should be >= 0")
```

Comment:
Why `-1`? Given that those are HTTP status codes, a value of `-1` would be extremely weird.

Why not use `400 Bad Request` instead?

According to the [RFC](https://tools.ietf.org/html/rfc7231#section-6.5.1):
> The 400 (Bad Request) status code indicates that the server cannot or will not process the
> request due to something that is perceived to be a client error (e.g., malformed request
> syntax, invalid request message framing, or deceptive request routing).

Same in other places where `-1` is used.

Reply:
The other return values are ignored (as usual) if the `error` is non-`nil`, so it's irrelevant what the first return value is.  (However, the [zero value](https://tour.golang.org/basics/12) might have been a better value to use.)

Reply:
Fixed to use zero values in #36